flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/RemoteStreamEnvironment.java

```diff
@@ -169,33 +173,66 @@ public RemoteStreamEnvironment(String host, int port, Configuration clientConfig
 		}
 	}
 
-	@Override
-	public JobExecutionResult execute(String jobName) throws ProgramInvocationException {
-		StreamGraph streamGraph = getStreamGraph();
-		streamGraph.setJobName(jobName);
-		transformations.clear();
-		return executeRemotely(streamGraph, jarFiles);
+	/**
+	 * Set savepoint restore settings that will be used when executing the job.
+	 * @param savepointRestoreSettings savepoint restore settings
+	 */
+	public void setSavepointRestoreSettings(SavepointRestoreSettings savepointRestoreSettings) {
+		this.savepointRestoreSettings = savepointRestoreSettings;
```

Comment:
Could we remove this setter? Savepoint restore should be configured per-execution basis and not for the lifetime of the `RemoteStreamEnvironment`.

It is not necessary because we have an addtional `executeRemotely` method now which takes the `SavepointRestoreSettings`. If you want to avoid too many parameters, add an additional non-static method which takes the savepoint settings.

Reply:
I second @mxm here, why not have e.g. additional method like: `execute(String jobName, SavepointSettings savepointSettings)` that a user that already have a `RemoteEnvironemt` could use.

So to sum up methods we would have:
* the actual submission
```
private static JobExecutionResult executeRemotely(StreamGraph streamGraph,
												ClassLoader envClassLoader,
												ExecutionConfig executionConfig,
												List<URL> jarFiles,
												String host,
												int port,
												Configuration clientConfiguration,
												List<URL> globalClasspaths,
												SavepointRestoreSettings savepointRestoreSettings
) throws ProgramInvocationException {
```

Also I would annotate all new methods that we introduce here as `@Internal` so that we can change them in future more easily.
* because we have to be backwards compatible
```
protected JobExecutionResult executeRemotely(StreamGraph streamGraph, List<URL> jarFiles) throws ProgramInvocationException {
```
* because we cannot change `StreamExecutionEnvironment:
```
@Override
public JobExecutionResult execute(String jobName) throws ProgramInvocationException {
```
* because we want to make it easier for users who have already `RemoteEnvironment`:
```
public JobExecutionResult execute(String jobName, SavepointRestoreSettings savepointRestoreSettings) throws ProgramInvocationException {
```
* because we want to be able to submit any `StreamEnvironment`:
```
public static JobExecutionResult executeRemotely(StreamExecutionEnvironment streamExecutionEnvironment,
	List<URL> jarFiles,
	String host,
	int port,
	Configuration clientConfiguration,
	List<URL> globalClasspaths,
	String jobName,
	SavepointRestoreSettings savepointRestoreSettings
) throws ProgramInvocationException {
```

Reply:
To be fair the code above has been changed, it's not a setter anymore but set in the constructor. You summed up quite nicely what is actually the state of this PR now. Nothing is broken in terms of backwards-compatibility but the structure changed a bit which makes it hard to review. The only difference is that an additional constructor has been added for the savepoint settings instead of this method you proposed:

```
public JobExecutionResult execute(String jobName, SavepointRestoreSettings savepointRestoreSettings) throws ProgramInvocationException {
```

Reply:
Unless you prefer the method instead of the constructor, then this should be ready to be merged :)

Reply:
Personally I do prefer the method. With the constructor your original concern still stands: 
>  Savepoint restore should be configured per-execution basis and not for the lifetime of the RemoteStreamEnvironment

Reply:
This argument was countered by the fact that the other parameters are also not per-execution. I don't mind, both the constructor or the method work for me.

Perhaps we could change the constructor to the method above @tweise? I think we should have reached consensus then :)